### ISSUE TITLE
[DFA-arch] Analytic streaming-bytes instrument — make the BPTT-memory claim of F19/F21 decidable instead of unmeasurable

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#159]
## Why

Two lanes now report the same negative on the same claim, and neither can
actually test it. toy#155 (F19) and toy#157 (F21) both state their success
target as *"DFA matches BP at k-times-less activation memory for sequence
length L"*, and both had to report the opposite of a win:

| lane | instrument | per-step cut vs BPTT |
|---|---|---|
| toy#155 | realized graph NODES | 31% MORE |
| toy#157 | realized graph BYTES | 17% MORE, and no better with L |

The reason is structural, not a bug: **in a graph autodiff every forward
tensor is materialised whatever the credit rule.** The detach dups and the
per-step surrogate terms are net additions, so the streaming win DFA promises
is not expressible as a graph rewrite. tao#21 records this and lists the fix
options; this is the cheapest of them, and the only one that does not require
a second execution engine.

## What to build

Alongside the MEASURED `graph: nodes=… bytes=…` line, emit an **ANALYTIC**
figure derived from the model's own shapes — the peak activation bytes a
STREAMING implementation would hold — for each arm:

```
graph:  nodes=2116 bytes=15578112          # what THIS harness builds (measured)
stream: bptt=13289988 cut=243456 ratio=54.6x recompute=2.0x   # what a streaming impl would hold (analytic)
```

- `bptt` — every step's forward activations must stay live for the ordered
  backward: **O(T)**.
- `cut` — with no gradient crossing a timestep, each step's update is
  independent, so one step's activations need be live at a time: **O(1) in T**.
- The instrument must be honest about what buys `cut`, because it is not free:
  the surrogate's error `e = softmax(logits) − labels` is only known after the
  readout at `t = T`, so an O(1) implementation must **recompute** the forward
  at update time. That is legitimate and is exactly what the absence of a
  time-chained backward permits — but it costs ~2x forward compute and the
  number must carry that (`recompute=2.0x`), or it is a rigged comparison.
- For the same reason it must also state BPTT's own best counter-move:
  sqrt-T gradient checkpointing gets BPTT to **O(√T)** at ~1.5x compute. The
  honest headline is the cut vs BOTH, not vs naive BPTT alone.

Derive it from the cell's live-set (per layer, per step: the 4 gate
pre-activations and post-activations, `c_t`, `tanh(c_t)`, `h_t`, plus `x_t`)
— **not** by walking the graph. Walking the graph is what already failed;
the point of this instrument is to answer "what would a streaming
implementation hold", which this harness never builds.

## Scope

- `lstm` first (toy#157 is where the target is stated in bytes), then mirror
  onto `ssm` so toy#155's table can be restated in the same units.
- Gate leg: assert `cut` is O(1) in T (equal across `--seq 16/64/128`) while
  `bptt` is linear in it, and that the MEASURED graph bytes still contradict
  both — the gap between the two instruments is the finding, and it must not
  quietly close.
- Docs: `docs/cli.md` (the `stream:` line, and that it is analytic, not
  measured) + the toy#155/#157 roadmap sections restated with it.

## What this does NOT do

It does not make toy stream. It replaces "we cannot measure the ticket's
target" with "here is the target's value under a stated model, next to the
measured cost of not implementing it" — enough for F19/F21 to be decided
rather than deferred, and enough to confirm or kill tao#21's recommendation
against spending the deferred F19 CUDA twin (a device port changes
throughput, not what the graph materialises).

Refs: tao#21, toy#155, toy#157.

**Comment by OriPekelman** [comment]
## Shipped in `6c0b8f7` — and the instrument sharpened the claim rather than just quantifying it

Both recurrent lanes now print the analytic line beside the measured one:

```
graph:  nodes=2116 bytes=15578112                       # MEASURED — what toy builds
stream: bptt=3868160 sqrt_t=984576 cut=78336 cut_vs_bptt=49.37x cut_vs_sqrt_t=12.56x replay=2x_fwd
```

**LSTM lane** (hidden 64, batch 32, 1 layer, 4 classes), bytes:

| T | bptt | sqrt_t | **cut** |
|---|---|---|---|
| 16 | 968 192 | 501 248 | **78 336** |
| 64 | 3 868 160 | 984 576 | **78 336** |
| 128 | 7 734 784 | 1 467 904 | **78 336** |
| 1024 | 61 867 520 | 3 884 544 | **78 336** |

**SSM lane** (d_inner 48, d_model 24, batch 32, 2 layers, K=4): cut = 213 504 flat, bptt 2 606 592 → 20 841 984 over T=16→128. `--selection lti` reports a strictly smaller cell (127 488) because its dt/C/gate branches are weights, not per-step activations — gated, since equal figures would mean the instrument was not reading the selection.

Constant in T, which is the claim: **49x under naive BPTT at T=64, 790x at T=1024, and still 12.6x / 49.6x against sqrt-T checkpointing** — BP's own best counter-move, quoted so the win is not measured only against BP's worst case.

### The part the ticket did not anticipate

The win is **not a property of the credit rule alone.** The surrogate's error `e = softmax(logits) − labels` is known only after the last-step readout, and step t's update needs step t's own activations. So the cut has exactly two ways to spend itself:

- **keep** every step's activations until the error arrives → O(T), which is what this harness does, and precisely why the *measured* graph is bigger; or
- **replay** the forward once the error is known, updating each step as it passes → O(1), at ~2x forward compute.

Replay is legal *because* no gradient crosses a timestep — the steps are independent, so they can be revisited in any order. Under BPTT it is not available at all: the backward must traverse time in order. **What cutting the time axis buys is the LEGALITY of the replay, not a smaller graph.** That reframes tao#21: the two instruments are not in conflict, they answer different questions, and the gap between them is the finding. Both ship, with the compute price inline (`replay=2x_fwd`) so the number cannot be quoted as free.

### Gated (both lanes)

- `cut` byte-identical across `--seq 16/64/128` — O(1) is asserted, not asserted-by-comment.
- `bptt` grows ~4x from T=16 to T=64.
- `cut` below `sqrt_t`.
- the MEASURED graph bytes still exceed the analytic figure — if that gap ever closed, toy started streaming, which is a far bigger change than this instrument.
- ssm additionally: the causal conv's K−1 window stays in the CARRY (O(K), not O(T)), and lti < selective.

Counted: per-step tensors a backward needs, per layer, excluding pre-activations whose derivative is computable from the output. Not counted: weights, optimizer moments, gradient buffers — equal across arms and independent of T, so they would only dilute the ratio. Full enumeration against each engine is in `lib/toy/train/stream_bytes.rb`.

`GATE PASS [lstm]` (12 legs), `GATE PASS [ssm]` (11 legs), `GATE PASS [train-cli-matrix]`. Docs: `docs/cli.md` §`stream:` and the toy#159 section of `docs/roadmap/dfa-arch-program-2026-08-10.md`. toy#160 can consume the line as specified.

**Comment by OriPekelman** [comment]
Closing: shipped in `6c0b8f7` — feat(dfa-arch): toy#159 analytic streaming-bytes instrument — the memory claim becomes decidable; revert toy#157's warmup default

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
